lib/matplotlib/__init__.py

```diff
@@ -1515,6 +1517,281 @@ def test(verbosity=1):
+                        label_namer=None, positional_parameter_names=None):
+    """
+    A decorator to add a 'data' kwarg to any a function.  The signature
+    of the input function must include the ax argument at the first position ::
+
+       def foo(ax, *args, **kwargs)
+
+    so this is suitable for use with Axes methods.
+
+    Parameters
+    ----------
+    replace_names : list of strings, optional, default: None
+        The list of parameter names which arguments should be replaced by
+        `data[name]`. If None, all arguments are replaced if they are
+        included in `data`.
+    replace_all_args : bool, default: False
+        If True, all arguments in *args get replaced, even if they are not
+        in replace_names.
+    label_namer : string, optional, default: None
+        The name of the parameter which argument should be used as label, if
+        label is not set. If None, the label keyword argument is not set.
+    positional_parameter_names : list of strings or callable, optional
+        The full list of positional parameter names (excluding an explicit
+        `ax`/'self' argument at the first place and including all possible
+        positional parameter in `*args`), in the right order. Can also include
+        all other keyword parameter. Only needed if the wrapped function does
+        contain `*args` and (replace_names is not None or replace_all_args is
+        False). If it is a callable, it will be called with the actual
+        tuple of *args and the data and should return a list like
+        above.
+        NOTE: callables should only be used when the names and order of *args
+        can only be determined at runtime. Please use list of names
+        when the order and names of *args is clear before runtime!
+    """
+    if replace_names is not None:
+        replace_names = set(replace_names)
+
+    def param(func):
+        new_sig = None
+        ver_info = sys.version_info
+        # py2.6 compatible version, use line below as soon as we can
+        _python_has_signature = ver_info[0] > 2 and ver_info[1] > 2
+        # _python_has_signature = ver_info.major > 2 and ver_info.minor > 2
+        if not _python_has_signature:
+            arg_spec = inspect.getargspec(func)
+            _arg_names = arg_spec.args
+            _has_no_varargs = arg_spec.varargs is None
+            _has_varkwargs = arg_spec.keywords is not None
+        else:
+            sig = inspect.signature(func)
+            _has_no_varargs = True
+            _has_varkwargs = False
+            _arg_names = []
+            params = list(sig.parameters.values())
+            for p in params:
+                if p.kind is p.VAR_POSITIONAL:
+                    _has_no_varargs = False
+                elif p.kind is p.VAR_KEYWORD:
+                    _has_varkwargs = True
+                else:
+                    _arg_names.append(p.name)
+            data_param = inspect.Parameter('data',
```

Comment:
I interprete that that `__signature__` is responsible for the inline code completition. But shouldn't boilerplate the do something similar (assuming that it doesn't use `__signature__` but the arg spec, which usually doesn't have `data` yet as it comes from the wrapped method?)